Implement async support for open_datatree

aladinor opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [X] Closes #10579 and [#12](https://github.com/aladinor/raw2zarr/issues/12) 
- [X] Tests added
- [X] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

aladinor committed:
adding async for datatrees

aladinor committed:
adding async method to _maybe_create_index

aladinor committed:
using async as complete instead of gathering results

aladinor committed:
adding tests for open_group, open_dtree and _maybe_create_index using async

aladinor committed:
ensuing _maybe_create_default_indexes_async is compatible with zarr v2

aladinor committed:
resolving the mypy type errors

aladinor committed:
attemp 2: resolving mypy type errors

shoyer commented:
This looks great! Would it be possible to make the sync path reuse the async methods internally? This would help reduce duplication, increase test coverage and speed up sync workflows.

aladinor committed:
refactor: consolidate async index creation for DataTree opening

Changes:
- Refactor open_datatree() to use zarr_sync() with async implementation
  for concurrent dataset and index creation across groups
- Add _open_datatree_from_stores_async() helper that opens datasets and
  creates indexes concurrently using asyncio.gather with a semaphore
  to limit concurrency (avoids deadlocks with stores like Icechunk)
- Add open_datatree_async() method for explicit async API
- Remove duplicate _maybe_create_default_indexes_async from zarr.py,
  now imports from api.py (single source of truth)

This significantly improves performance when opening DataTrees from
high-latency storage backends (e.g., ~2 seconds vs sequential loading).

aladinor committed:
perf: remove unnecessary semaphore from async datatree opening

Remove the asyncio.Semaphore that was limiting concurrency to 10
concurrent operations. Investigation showed:

- Zarr already has built-in concurrency control (async.concurrency=10)
- The semaphore only applied to asyncio.to_thread() calls, not zarr I/O
- Removing it improves performance by ~30-40% (~2s -> ~1.2-1.4s)

The semaphore was defensive code for a problem that doesn't exist -
zarr and icechunk handle their own concurrency limits internally.

aladinor committed:
fix: add zarr v2 fallback for datatree opening

The async implementation uses zarr.core.sync which only exists in
zarr v3. Add a conditional check using _zarr_v3() to:
- Use async path with zarr_sync() for zarr v3 (concurrent loading)
- Fall back to sequential loading for zarr v2

This fixes CI failures on min-versions environment which uses zarr v2.

aladinor committed:
updating whats-new.rst file

aladinor committed:
fix: re-add semaphore to async datatree opening to prevent deadlocks

The previous commit (0ee2a73f) removed the semaphore thinking zarr handles
its own concurrency, but icechunk can deadlock when too many asyncio.to_thread()
calls try to access it simultaneously. This was discovered when testing with
larger stores (23+ groups) where all threads would start but never complete.

The semaphore limits concurrent to_thread calls to 10, which prevents the
deadlock while still providing significant performance benefits over sequential
loading.

aladinor committed:
refactor: use async index creation in sync open_datatree for zarr

- Add helper methods _build_group_members and _create_stores_from_members
  to reduce code duplication between sync and async store opening
- Use zarr_sync() to run async index creation in _datatree_from_backend_datatree
  for zarr engine, making open_datatree fully async behind the scenes
- Fix missing chunks validation and source encoding in open_datatree_async
- Add tests for chunks validation, source encoding, and chunks parameter

aladinor committed:
fix: add type ignore for mypy arg-type error in open_datatree_async

aladinor committed:
fix: add type annotations and fix Windows path in test

- Add type annotations to nested async functions in _datatree_from_backend_datatree
  to fix mypy annotation-unchecked notes breaking pytest-mypy-plugins tests
- Use os.path.join and os.path.normpath in test_async_source_encoding
  for cross-platform compatibility on Windows

aladinor committed:
fix: add type annotations to nested async functions for mypy

Add type annotations to _maybe_create_default_indexes_async and its
nested functions (load_var, create_index, _create) to satisfy mypy's
annotation-unchecked checks. Also add Variable and Hashable imports
to the TYPE_CHECKING block.

This fixes pytest-mypy-plugins tests that were failing due to mypy
emitting annotation-unchecked notes for untyped nested functions.

aladinor committed:
refactor: remove public open_datatree_async API per review feedback

- Remove open_datatree_async() from api.py (public API)
- Remove open_datatree_async() from zarr.py (backend method)
- Keep internal async optimization in _datatree_from_backend_datatree()
- Use _zarr_v3() for proper zarr version check instead of ImportError
- Update tests to only test internal async functionality
- Add test to verify sync open_datatree uses async internally for zarr v3

The async optimization is now internal only - users call the sync
open_datatree() which automatically uses async index creation for
zarr v3 backends.

keewis commented on xarray/backends/zarr.py:
```diff
@@ -1709,25 +1908,122 @@ def open_datatree(
+
+        # Limit concurrent to_thread calls to avoid deadlocks with some stores
+        # (e.g., icechunk can deadlock when too many threads access it simultaneously)
+        sem = asyncio.Semaphore(10)
```
not sure if this is necessary right now, but we might need to make the number of concurrent `to_thread` class a parameter

keewis replied:
I would have considered this a knob for fine-tuning I/O performance, not something a user would have to interact with. Either way, I agree that this is something we can safely postpone to a later PR.

aladinor committed:
refactor: convert _build_group_members to module-level helper function

aladinor committed:
fix: add cast for mypy type checking in _build_group_members

keewis commented on xarray/backends/api.py:
```diff
@@ -385,6 +417,59 @@ def _datatree_from_backend_datatree(
+        for var_name in idx_vars:
+            indexes[var_name] = idx
+
+    new_coords = Coordinates._construct_direct(coords=variables, indexes=indexes)
```
I wonder if we should move the creation of the index out of this function? As far as I understand, the performance bottleneck is solved by fetching as many coordinates as possible concurrently in the `asyncio.gather` call above, but the creation of the index is CPU-bound so doesn't really benefit from async.

That would allow sharing code paths with the sync code, which basically just pulls out the 1D dimension variables and has the `Coordinates` constructor create the indexes.

I guess the downside would be that we'd be iterating over the hierarchy twice.

keewis reviewed (approved):
I'm not an expert on the zarr backend, but this mostly looks good to me.

This is not done for the sync versions either, so I don't think this has to be done in this PR, but _logically_ I think `xr.open_datatree(...) == xr.DataTree.from_dict(xr.open_groups(...))`, so it might make sense to have `_open_datatree_async` call `_open_groups_as_dict_async`?

aladinor committed:
Update xarray/backends/api.py

Co-authored-by: Justus Magin <keewis@users.noreply.github.com>

aladinor committed:
refactor: use sync index creation in _maybe_create_default_indexes_async

Benchmarking showed async index creation provides no measurable benefit
since it's CPU-bound work. Simplified to sync loop per reviewer feedback.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

aladinor commented:
@keewis, you're right that open_datatree ≈ DataTree.from_dict(open_groups_as_dict(...)) and refactoring to share code would reduce duplication. Since this affects both sync and async paths, I'll address it in a follow-up PR to keep this one focused. I've noted the differences (semaphore, index creation) that need to be unified.

> I'm not an expert on the zarr backend, but this mostly looks good to me.
> 
> This is not done for the sync versions either, so I don't think this has to be done in this PR, but _logically_ I think `xr.open_datatree(...) == xr.DataTree.from_dict(xr.open_groups(...))`, so it might make sense to have `_open_datatree_async` call `_open_groups_as_dict_async`?

keewis commented on xarray/backends/api.py:
```diff
@@ -446,23 +446,16 @@ async def load_var(var: Variable) -> Variable:
         indexes[name] = idx
         variables.update(idx.create_variables({name: ds.variables[name]}))
+
+    new_coords = Coordinates._construct_direct(coords=variables, indexes=indexes)
```
since the creation of the indexes are sync now, it may be better to just use `xr.Coordinates`:
```suggestion
    variables = {name: ds.variables[name] for name in to_index_names}
    new_coords = Coordinates(variables)
```

aladinor committed:
Update xarray/backends/api.py

Co-authored-by: Justus Magin <keewis@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

aladinor committed:
Update xarray/backends/api.py

Co-authored-by: Justus Magin <keewis@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

aladinor committed:
Address PR review: TaskGroup, max_concurrency, and open_dataset_async

- Replace asyncio.gather with asyncio.TaskGroup for better error handling
  (cancels outstanding tasks on error)
- Add max_concurrency parameter to open_datatree for controlling parallel
  I/O operations (defaults to 10)
- Add StoreBackendEntrypoint.open_dataset_async method
- Add test for open_dataset_async equivalence

aladinor committed:
Fix async deadlock risks: use zarr built-in members(), run_in_executor safety

- Replace custom _iter_zarr_groups_async (~90 lines) with zarr's
  AsyncGroup.members(max_depth=None) to avoid sync fallback deadlock
- Wrap _get_open_params and _build_group_members in run_in_executor
  in open_store_async to prevent reentrant sync() deadlock
- Add dedicated executor to _maybe_create_default_indexes_async and
  create_indexes_async to avoid thread pool exhaustion on zarr's IO loop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

aladinor committed:
Fix zarr v2 fallback in _iter_zarr_groups_async

Add try/except around zarr.core.group import so zarr v2 (where
zarr.core is a module, not a package) falls back to sync discovery.

aladinor committed:
Trim verbose comments to match xarray style

aladinor committed:
Remove double index creation from async DataTree open

The zarr v3 async backend was creating indexes inside
_open_datatree_from_stores_async, but the framework layer
(_datatree_from_backend_datatree in api.py) already handles
index creation for all backends. This caused indexes to be
created twice per node and made create_default_indexes=False
silently ignored.

Remove the index creation from the backend to restore the
clean main architecture: backends open datasets without
indexes, the framework creates them once.

aladinor committed:
Use native async for DataTree open, threads only for CPU decode

Replace the per-group thread+zarr_sync pattern in
_open_datatree_from_stores_async with a two-phase approach:

Phase 1: Single async_root.members(max_depth=None) call discovers
all groups AND their array members in one pass, replacing both
_iter_zarr_groups_async and per-group _fetch_members calls.

Phase 2: Wrap AsyncArray/AsyncGroup in sync Array/Group (zero-cost),
inject pre-fetched members into ZarrStore, run only CPU-bound
decode_cf_variables in thread pool.

Results (laptop, 60-node OSN store):
- zarr_sync calls: 122 → 2
- members() calls: 61 → 1
- ~15-20% faster open_datatree (2.08s → 1.77s with indexes)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

aladinor committed:
Merge branch 'main' into async-dtreec

aladinor committed:
Merge branch 'main' into async-dtreec

aladinor committed:
Merge branch 'main' into async-dtreec

aladinor committed:
Merge branch 'main' into async-dtreec

aladinor committed:
Merge branch 'main' into async-dtreec

aladinor committed:
Merge branch 'main' into async-dtreec
